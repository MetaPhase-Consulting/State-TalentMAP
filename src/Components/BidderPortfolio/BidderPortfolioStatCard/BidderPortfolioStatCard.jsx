import { useEffect, useState } from 'react';
import { get } from 'lodash';
import { Link } from 'react-router-dom';
import { Tooltip } from 'react-tippy';
import { Cusp, Eligible } from 'Components/Ribbon';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import FA from 'react-fontawesome';
import { BIDDER_OBJECT, CLASSIFICATIONS } from 'Constants/PropTypes';
import { NO_GRADE, NO_LANGUAGE, NO_POST, NO_TOUR_END_DATE } from 'Constants/SystemMessages';
import { formatDate } from 'utilities';
import TextareaAutosize from 'react-textarea-autosize';
import { saveBidderPortfolioSelections } from 'actions/bidderPortfolio';
import ToggleButton from 'Components/ToggleButton';
import InteractiveElement from 'Components/InteractiveElement';
import BoxShadow from '../../BoxShadow';
import SkillCodeList from '../../SkillCodeList';
import ClientBadgeList from '../ClientBadgeList';
import SearchAsClientButton from '../SearchAsClientButton';
import AddToInternalListButton from '../AddToInternalListButton';

const BidderPortfolioStatCard = ({ userProfile, showEdit, classifications }) => {
  const dispatch = useDispatch();
  const currentAssignmentText = get(userProfile, 'pos_location');
  const clientClassifications = get(userProfile, 'classifications');
  const perdet = get(userProfile, 'perdet_seq_number');
  const id = get(userProfile, 'employee_id');
  const ted = formatDate(get(userProfile, 'current_assignment.end_date'));
  const languages = get(userProfile, 'current_assignment.position.language');
  const bidder = get(userProfile, 'shortened_name') || 'None listed';
  // This is the new key bidder_types. It returns a string of either 'cusp' or 'eligible'
  const bidderType = get(userProfile, 'bidder_types') || null;
  const email = get(userProfile, 'cdos')[0]?.cdo_email || 'None listed';
  const orgShortDesc = get(userProfile, 'current_assignment.position.organization');
  const [currentBidderType, setCurrentBidderType] = useState(bidderType);
  const [included, setIncluded] = useState(bidderType === 'cusp');
  const [showMore, setShowMore] = useState(false);
  const [edit, setEdit] = useState(false);
  const [comments, setComments] = useState('');
  const [altEmail, setAltEmail] = useState('');
  const [verifyComments, setVerifyComments] = useState('');
  const [verifyAltEmail, setVerifyAltEmail] = useState('');

  const cusp = included;
  const eligible = !included;
  const showToggle = bidderType !== null;
  const showSaveAndCancel = edit && showMore;

  const editClient = (e) => {
    e.preventDefault();
    setEdit(previous => !previous);
  };

  const saveEdit = () => {
    setComments(verifyComments);
    setAltEmail(verifyAltEmail);
    // Nothing to do yet, will add later
    const clientData = {
      id,
      verifyComments,
      verifyAltEmail,
    };
    dispatch(saveBidderPortfolioSelections(clientData));
    setEdit(false);
  };

  useEffect(() => {
    if (currentBidderType === 'eligible') {
      setIncluded(false);
    }
    if (currentBidderType === 'cusp') {
      setIncluded(true);
    }
  }, [currentBidderType]);

  const onToggleChange = () => {
    console.log(currentBidderType);
    if (currentBidderType === 'cusp') {
      setCurrentBidderType('eligible');
    }
    if (currentBidderType === 'eligible') {
      setCurrentBidderType('cusp');
    }
  };

  const showSearch = !showEdit && !edit;
  const collapseCard = () => {
    setShowMore(!showMore);
    setEdit(false);
  };

  const onCancel = () => {
    setVerifyComments('');
    setVerifyAltEmail('');
    setEdit(false);
  };

  const ribbons = (
    <div>
      {
        eligible &&
        <Tooltip
          title="Eligible"
          arrow
          offset={-60}
        >
          <Eligible />
        </Tooltip>
      }
      {
        cusp &&
        <Tooltip
          title="Cusp"
          arrow
          offset={-60}
        >
          <Cusp />
        </Tooltip>
      }
    </div>
  );

  return (
    <BoxShadow className="usa-grid-full bidder-portfolio-stat-card">
      <div className="bidder-portfolio-stat-card-top">
        <div className="bidder-compact-card-head">
          <h3 className="stat-card-client">Client Overview</h3>
          {showToggle &&
            <ToggleButton
              labelTextRight={!included ? 'Excluded' : 'Included'}
              checked={included}
              onChange={onToggleChange}
              onColor="#0071BC"
            />
          }
        </div>
        <div className="stat-card-data-point bidder-compact-card-head">
          <Link to={`/profile/public/${perdet}`}>{bidder}</Link>
          { showMore &&
            <Link to="#" onClick={(e) => editClient(e)}>
              <FA name="pencil" />
                Edit
            </Link>
          }
        </div>
        {showToggle &&
          <div className="bidder-portfolio-ribbon-container">
            <div className="ribbon-container-condensed-min">
              {ribbons}
            </div>
          </div>
        }
        <div className="stat-card-data-point">
          <dt>Employee ID:</dt><dd>{id}</dd>
        </div>
        <div className="stat-card-data-point">
          <dt>Skill:</dt><dd><SkillCodeList skillCodes={userProfile.skills} /></dd>
        </div>
        <div className="stat-card-data-point">
          <dt>Grade:</dt><dd>{userProfile.grade || NO_GRADE}</dd>
        </div>
        <div className="stat-card-data-point">
          <dt>Languages:</dt><dd>{languages || NO_LANGUAGE}</dd>
        </div>
        <div className="stat-card-data-point">
          <dt>TED:</dt><dd>{ted || NO_TOUR_END_DATE}</dd>
        </div>
        <div className="stat-card-data-point">
          <dt className="location-label">Location (Org):</dt><dd>{currentAssignmentText || NO_POST} ({orgShortDesc})</dd>
        </div>
        <div className="stat-card-data-point">
          <dt>DOS Email:</dt>
          <dd>
            <a href={`mailto: ${email}`}>{email}</a>
          </dd>
        </div>
        <div className={!edit && 'stat-card-data-point'} >
          <dt>Alt Email:</dt>
          <dd>
            {altEmail ?
              <a href={`mailto:${altEmail}`}>{altEmail}</a> :
              'None Listed'
            }
          </dd>
          {edit &&
            <input
              type="text"
              defaultValue=""
              placeholder="example@gmail.com"
              onChange={(e) => setVerifyAltEmail(e.target.value)}
            />
          }
        </div>
      </div>
      <div className="bidder-portfolio-stat-card-bottom">
        <div>
          <span className="updates">Classifications: </span>
          <ClientBadgeList
            clientClassifications={clientClassifications}
            classifications={classifications}
          />
        </div>
        { showSearch &&
          <div className="button-container" style={{ display: 'flex', flexDirection: 'row', justifyContent: 'space-between' }}>
            <SearchAsClientButton user={userProfile} />
            <AddToInternalListButton refKey={perdet} />
          </div>
        }
      </div>
      <div className="bidder-portfolio-stat-card-bottom">
        { showMore &&
          <div>
            <dt>Comments:</dt>
            <div className="stat-card-data-point stat-card-comments">
              <TextareaAutosize
                className="stat-card-textarea"
                disabled={!edit}
                maxLength="255"
                name="note"
                placeholder="No Notes"
                defaultValue={!comments ? '' : comments}
                onChange={(e) => setVerifyComments(e.target.value)}
              />
            </div>
          </div>
        }
        { showSaveAndCancel &&
          <div className="stat-card-btn-container">
            <button className="stat-card-cancel-btn" onClick={onCancel}>Cancel</button>
            <button onClick={saveEdit} disabled={!verifyComments && !verifyAltEmail}>Save</button>
          </div>
        }
        <div className="toggle-more-container">
          <InteractiveElement className="toggle-more" onClick={collapseCard}>
            <FA name={`chevron-${showMore ? 'up' : 'down'}`} />
          </InteractiveElement>
        </div>
      </div>
    </BoxShadow>
  );
};

BidderPortfolioStatCard.propTypes = {
  userProfile: BIDDER_OBJECT.isRequired,
  showEdit: PropTypes.bool,
  classifications: CLASSIFICATIONS,
};

BidderPortfolioStatCard.defaultProps = {
  showEdit: false,
  classifications: [],
};

export default BidderPortfolioStatCard;
