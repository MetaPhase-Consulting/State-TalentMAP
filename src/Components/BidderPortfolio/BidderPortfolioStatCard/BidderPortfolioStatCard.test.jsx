import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import toJSON from 'enzyme-to-json';
import thunk from 'redux-thunk';
import BidderPortfolioStatCard from './BidderPortfolioStatCard';
import { clientObject } from '../../../__mocks__/client';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('BidderPortfolioStatCard', () => {
  const props = {
    userProfile: clientObject,
  };
  it('is defined', () => {
    const wrapper = shallow(
      <Provider store={mockStore({})}>
        <BidderPortfolioStatCard
          {...props}
        />
      </Provider>, 
      );
    expect(wrapper).toBeDefined();
  });

  it('matches snapshot', () => {
    const wrapper = shallow(
      <Provider store={mockStore({})}>
        <BidderPortfolioStatCard
          {...props}
        />
      </Provider>, 
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
