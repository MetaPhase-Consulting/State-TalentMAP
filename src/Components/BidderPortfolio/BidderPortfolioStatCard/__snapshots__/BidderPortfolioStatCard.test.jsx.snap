// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BidderPortfolioStatCard matches snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <BidderPortfolioStatCard
    classifications={Array []}
    showEdit={false}
    userProfile={
      Object {
        "assignments": Array [
          Object {
            "arrival_date": null,
            "bid_approval_date": "1975-01-01T00:00:00Z",
            "combined_differential": 0,
            "create_date": "2018-01-12T15:11:37.521934Z",
            "curtailment_reason": null,
            "end_date": null,
            "estimated_end_date": "2019-01-12T00:00:00Z",
            "id": 5,
            "is_domestic": false,
            "position": Object {
              "bureau": "(WHA) BUREAU OF WESTERN HEMISPHERIC AFFAIRS",
              "position_number": "58312265",
              "post": Object {
                "location": Object {
                  "city": "Freetown",
                  "code": "00A",
                  "country": "Sierra Leone",
                  "id": 103,
                  "state": "",
                },
              },
              "skill": "CONSTRUCTION ENGINEERING (6218)",
              "title": "PROJECT DIRECTOR",
            },
            "service_duration": null,
            "standard_tod_months": 0,
            "start_date": "2018-01-12T00:00:00Z",
            "status": "active",
            "tour_of_duty": "1 YR (3 R & R)",
            "update_date": "2018-01-12T15:12:01.013500Z",
            "user": "Kara Batisak",
          },
        ],
        "bid_statistics": Array [
          Object {
            "approved": 0,
            "bidcycle": "Demo BidCycle 2018-01-12 15:11:38.004793",
            "closed": 0,
            "declined": 0,
            "draft": 5,
            "handshake_accepted": 0,
            "handshake_declined": 0,
            "handshake_offered": 3,
            "id": 1,
            "in_panel": 0,
            "submitted": 2,
            "user": "Kara Batisak",
          },
        ],
        "current_assignment": Object {
          "arrival_date": null,
          "bid_approval_date": "1975-01-01T00:00:00Z",
          "combined_differential": 0,
          "create_date": "2018-01-12T15:11:37.521934Z",
          "curtailment_reason": null,
          "end_date": null,
          "estimated_end_date": "2019-01-12T00:00:00Z",
          "id": 5,
          "is_domestic": false,
          "position": Object {
            "bureau": "(WHA) BUREAU OF WESTERN HEMISPHERIC AFFAIRS",
            "position_number": "58312265",
            "post": Object {
              "location": Object {
                "city": "Freetown",
                "code": "00A",
                "country": "Sierra Leone",
                "id": 103,
                "state": "",
              },
            },
            "skill": "CONSTRUCTION ENGINEERING (6218)",
            "title": "PROJECT DIRECTOR",
          },
          "service_duration": null,
          "standard_tod_months": 0,
          "start_date": "2018-01-12T00:00:00Z",
          "status": "active",
          "tour_of_duty": "1 YR (3 R & R)",
          "update_date": "2018-01-12T15:12:01.013500Z",
          "user": "Kara Batisak",
        },
        "grade": "01",
        "id": 5,
        "is_cdo": false,
        "language_qualifications": Array [],
        "primary_nationality": "United States",
        "secondary_nationality": null,
        "skills": Array [
          Object {
            "code": "6218",
            "cone": null,
            "description": "CONSTRUCTION ENGINEERING",
            "id": 63,
          },
        ],
        "user": Object {
          "email": "batisak@state.gov",
          "first_name": "Kara",
          "last_name": "Batisak",
          "username": "batisak",
        },
      }
    }
  />
</ContextProvider>
`;
